tests/lax_control_flow_test.py

```diff
@@ -2050,29 +2106,54 @@ def sqrt_cubed(x, tangent_solve=scalar_solve):
     self.assertAllClose(
         results, 5.0**1.5, check_dtypes=False, rtol={np.float64: 1e-7})
 
+  @parameterized.named_parameters([
+    {"testcase_name": "vector_solve", "solve_method": vector_solve},
+    {"testcase_name": "newton_raphson", "solve_method": newton_raphson},
+  ])
   @jtu.skip_on_flag("jax_skip_slow_tests", True)
-  def test_custom_root_vector_with_solve_closure(self):
-
-    def vector_solve(f, y):
-      return jnp.linalg.solve(jax.jacobian(f)(y), y)
+  def test_custom_root_vector_with_solve_closure(self, solve_method):
 
     def linear_solve(a, b):
       f = lambda y: high_precision_dot(a, y) - b
       x0 = jnp.zeros_like(b)
       solution = jnp.linalg.solve(a, b)
       oracle = lambda func, x0: solution
-      return lax.custom_root(f, x0, oracle, vector_solve)
+      return lax.custom_root(f, x0, oracle, solve_method)
 
     rng = np.random.RandomState(0)
     a = rng.randn(2, 2)
     b = rng.randn(2)
     jtu.check_grads(linear_solve, (a, b), order=2,
-                    atol={np.float32: 1e-2, np.float64: 1e-11})
+                    atol={np.float32: 1e-2, np.float64: 1e-11},
+                    modes=["fwd"],
```

Comment:
Does `rev` mode no longer work? If so, that would be a concern...

Reply:
Spent some time looking in to this. There's no regression introduced here, but the behavior is definitely not ideal. To be explicit, here's the status of the `fwd` and `rev` mode:

 - For `vector_solve`, both `fwd` and `rev` work, as before. I'll make sure to add the `rev` mode back for that function.
 - For `newton_raphson`, only `fwd` mode works with vectors in this test (but see below about scalars vs. vectors).

The key problem is that JAX traces the function during reverse mode differentiation, and it hits a `while_loop`, which is not reverse-mode differentiable. I've verified that it's a problem in the original code as well—for example, you can replace `vector_solve` with this equivalent function and trigger the same error:

```python
def vector_solve(f, y):
  # This while_loop is an identity operation on y:
  y = lax.while_loop(lambda _: False, lambda _: _, y)
  return jnp.linalg.solve(jax.jacobian(f)(y), y)
```

I believe that the JVP definition provides all of the necessary information for the VJP, so tracing seems completely unnecessary. Nevertheless, I'm having difficulty tracking down where and why the tracer attempts to trace in `rev` mode. 

My suspicion is that reverse mode uses tracing as a shortcut for verifying array shapes.


### Scalars vs. Vectors

If we replace `y` in [line 2150](https://github.com/google/jax/pull/8869/files#diff-c3f68aab0e05d9d2de62fd88802c96c19d9060f6d2500a7eb4266ee4139e717cR2150) with a scalar, then reverse mode works on the newton method as well.

I suspect that there is some optimization that shortcuts tracing in the case of a scalar. This would explain why both `binary_search` and `newton_rasphson` are reverse-mode differentiable in `test_custom_root_scalar` despite their `while_loop`.

Reply:
For completeness, the reverse mode test for newton_raphson fails with error:

```
E     ValueError: Reverse-mode differentiation does not work for lax.while_loop or lax.fori_loop. Try using lax.scan instead.
```